### PR TITLE
[Feature] Add ArcI Model.

### PR DESCRIPTION
**Pull request by pl8787**
[Feature] Add ArcI Model. for #311

### CODE REVIEW

File: matchzoo/models/arci_model.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    """
+
+    @classmethod
+    def get_default_params(cls) -> engine.ParamTable:
+        """:return: model default parameters."""
+        params = super().get_default_params()
+        params['optimizer'] = 'adam'
+        params['input_shapes'] = [(32,), (32,)]
+        params['input_dtypes'] = [np.int32, np.int32]
```
Comment (bwanglzu) [inline matchzoo/models/arci_model.py]: you defined `input_dtypes`, but never used in the model

File: matchzoo/models/arci_model.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        params.add(engine.Param('maxlen_left', 32))
+        params.add(engine.Param('maxlen_right', 32))
+        params.add(engine.Param('embed_finetune', False))
```
Comment (bwanglzu) [inline matchzoo/models/arci_model.py]: the variable name is ambiguous, `trainable` is better

File: matchzoo/models/arci_model.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        params.add(engine.Param('embed_dim', 300))
```
Comment (bwanglzu) [inline matchzoo/models/arci_model.py]: follow the same naming convension `dim_embedding`

File: matchzoo/models/arci_model.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        params.add(engine.Param('vocab_size', 100))
+        params.add(engine.Param('left_kernel_count', 32))
+        params.add(engine.Param('left_kernel_size', 3))
+        params.add(engine.Param('right_kernel_count', 32))
+        params.add(engine.Param('right_kernel_size', 3))
+        params.add(engine.Param('q_pool_size', 16))
+        params.add(engine.Param('d_pool_size', 16))
```
Comment (bwanglzu) [inline matchzoo/models/arci_model.py]: `q_pool_size` denotes for  `query_pool_size`? then why not `left_pool_size`? same for `d_pool_size`

File: matchzoo/models/arci_model.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        # Process left & right input.
+        embedding = Embedding(self._params['vocab_size'],
+                              self._params['embed_dim'],
+                              weights=[self._params['embed']],
+                              trainable=self._params['embed_finetune'])
+        embed_left = embedding(input_left)
+        embed_right = embedding(input_right)
+        conv_left = Conv1D(self._params['left_kernel_count'],
+                           self._params['left_kernel_size'],
+                           padding='same',
```
Comment (bwanglzu) [inline matchzoo/models/arci_model.py]: `padding = self._params['padding']`, make it tunable.

File: matchzoo/models/arci_model.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                           activation='relu')(embed_left)
```
Comment (bwanglzu) [inline matchzoo/models/arci_model.py]: `activation = self._params["activation"]`

File: matchzoo/models/arci_model.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        conv_right = Conv1D(self._params['right_kernel_count'],
+                            self._params['right_kernel_size'],
+                            padding='same',
```
Comment (bwanglzu) [inline matchzoo/models/arci_model.py]: same

File: matchzoo/models/arci_model.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                            activation='relu')(embed_right)
```
Comment (bwanglzu) [inline matchzoo/models/arci_model.py]: same

File: matchzoo/models/arci_model.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+        params.add(engine.Param('left_kernel_count', 32))
+        params.add(engine.Param('left_kernel_size', 3))
+        params.add(engine.Param('right_kernel_count', 32))
+        params.add(engine.Param('right_kernel_size', 3))
+        params.add(engine.Param('activation', 'relu'))
+        params.add(engine.Param('left_pool_size', 16))
+        params.add(engine.Param('right_pool_size', 16))
+        params.add(engine.Param('padding', 'same'))
+        params.add(engine.Param('dropout_rate', 0.0))
+        params.add(engine.Param('embed',
```
Comment (bwanglzu) [inline matchzoo/models/arci_model.py]: sometimes you're using `embedding` (line 27, 28), sometimes you're using `embed`

File: matchzoo/models/arci_model.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        params['optimizer'] = 'adam'
+        params['input_shapes'] = [(32,), (32,)]
+        params['input_dtypes'] = [np.int32, np.int32]
+        params.add(engine.Param('trainable_embedding', False))
+        params.add(engine.Param('dim_embedding', 300))
+        params.add(engine.Param('vocab_size', 100))
+        params.add(engine.Param('left_kernel_count', 32))
+        params.add(engine.Param('left_kernel_size', 3))
+        params.add(engine.Param('right_kernel_count', 32))
+        params.add(engine.Param('right_kernel_size', 3))
```
Comment (bwanglzu) [inline matchzoo/models/arci_model.py]: can we use `kernel_count` and `kernel_size` and remove the left and right?

Comment (pl8787) [reply matchzoo/models/arci_model.py]: In information retrieval task, they use different size.

File: matchzoo/models/arci_model.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        assert(params['num_blocks'] == len(params['left_kernel_size']))
+        assert(params['num_blocks'] == len(params['left_pool_size']))
+        assert(params['num_blocks'] == len(params['right_kernel_count']))
+        assert(params['num_blocks'] == len(params['right_kernel_size']))
+        assert(params['num_blocks'] == len(params['right_pool_size']))
+        return params
+
+    def _conv_pool_block(self, input, kernel_count: int, kernel_size: int,
+                         padding: str, activation: str, pool_size: int):
+        output = Conv1D(kernel_count, kernel_size,
```
Comment (bwanglzu) [inline matchzoo/models/arci_model.py]:
```python
out = Conv1D(filters=kernel_count,
                       .....=.....
                       .....=.....
                       activation=..)
```

File: matchzoo/models/arci_model.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        assert(params['num_blocks'] == len(params['left_kernel_count']))
+        assert(params['num_blocks'] == len(params['left_kernel_size']))
+        assert(params['num_blocks'] == len(params['left_pool_size']))
+        assert(params['num_blocks'] == len(params['right_kernel_count']))
+        assert(params['num_blocks'] == len(params['right_kernel_size']))
+        assert(params['num_blocks'] == len(params['right_pool_size']))
+        return params
+
+    def _conv_pool_block(self, input, kernel_count: int, kernel_size: int,
+                         padding: str, activation: str, pool_size: int):
```
Comment (bwanglzu) [inline matchzoo/models/arci_model.py]: isn't `kernel_count`, `kernel_size`, `padding`, `activation`, `pool_size`  can be get from `self._params`?

Comment (pl8787) [reply matchzoo/models/arci_model.py]: left and right have different parameters.

File: matchzoo/models/arci_model.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+        return output
+
+    def build(self):
+        """
+        Build model structure.
+
+        ArcI use Siamese arthitecture.
+        """
+        # Left input and right input.
+        input_left = Input(name='id_left',
```
Comment (faneshion) [inline matchzoo/models/arci_model.py]: Right now, the `Input` name has been changed to `text_left`.

File: matchzoo/models/arci_model.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+                           shape=self._params['input_shapes'][0])
+        input_right = Input(name='id_right',
```
Comment (faneshion) [inline matchzoo/models/arci_model.py]: Same as above.

**Pull request merged**